src/transport/MessageEndpoint.cpp

```diff
@@ -88,171 +95,163 @@ void MessageEndpoint::open(faabric::transport::MessageContext& context,
-            throw;
-        }
-    } else {
-        try {
-            auto res = this->socket->send(zmq::buffer(serialisedMsg, msgSize),
-                                          zmq::send_flags::none);
-            if (res != msgSize) {
-                SPDLOG_ERROR("Sent different bytes than expected (sent "
-                             "{}, expected {})",
-                             res.value_or(0),
-                             msgSize);
-                throw std::runtime_error("Error sending message");
-            }
-        } catch (zmq::error_t& e) {
-            SPDLOG_ERROR("Error sending message: {}", e.what());
-            throw;
-        }
+
+    if (this->socket == nullptr) {
+        throw std::runtime_error("Sending on an unopened socket");
     }
+
+    if (socketType == SocketType::PULL) {
+        throw std::runtime_error("Sending on a recv socket");
+    }
+
+    zmq::send_flags sendFlags =
+      more ? zmq::send_flags::sndmore : zmq::send_flags::none;
+
+    CATCH_ZMQ_ERR(
+      {
+          auto res =
+            this->socket->send(zmq::buffer(serialisedMsg, msgSize), sendFlags);
+          if (res != msgSize) {
+              SPDLOG_ERROR("Sent different bytes than expected (sent "
+                           "{}, expected {})",
+                           res.value_or(0),
+                           msgSize);
+              throw std::runtime_error("Error sending message");
+          }
+      },
+      "send")
 }
 
 // By passing the expected recv buffer size, we instrument zeromq to receive on
 // our provisioned buffer
 Message MessageEndpoint::recv(int size)
 {
     assert(tid == std::this_thread::get_id());
-    assert(this->socket != nullptr);
     assert(size >= 0);
 
-    // Pre-allocate buffer to avoid copying data
-    if (size > 0) {
-        Message msg(size);
-
-        try {
-            auto res = this->socket->recv(zmq::buffer(msg.udata(), msg.size()));
-
-            if (!res.has_value()) {
-                SPDLOG_ERROR("Timed out receiving message of size {}", size);
-                throw MessageTimeoutException("Timed out receiving message");
-            }
-
-            if (res.has_value() && (res->size != res->untruncated_size)) {
-                SPDLOG_ERROR("Received more bytes than buffer can hold. "
-                             "Received: {}, capacity {}",
-                             res->untruncated_size,
-                             res->size);
-                throw std::runtime_error("Error receiving message");
-            }
-        } catch (zmq::error_t& e) {
-            if (e.num() == ZMQ_ETERM) {
-                // Return empty message to signify termination
-                SPDLOG_TRACE("Shutting endpoint down after receiving ETERM");
-                return Message();
-            }
-
-            // Print default message and rethrow
-            SPDLOG_ERROR("Error receiving message: {} ({})", e.num(), e.what());
-            throw;
-        }
```

Comment:
To make things simpler to parse I split `recv` up into two methods based on whether a `size` is provided.